client.go

```diff
@@ -563,40 +611,57 @@ func (c *Client) handleLine(line string) error {
 	}
 
 	if strings.HasPrefix(line, "@") {
-		channel, user, clientMessage := ParseMessage(line)
+		message := parseMessage(line)
 
-		switch clientMessage.Type {
-		case PRIVMSG:
-			if c.onNewMessage != nil {
-				c.onNewMessage(channel, *user, *clientMessage)
-			}
+		switch message.RawMessage.Type {
 		case WHISPER:
 			if c.onNewWhisper != nil {
-				c.onNewWhisper(*user, *clientMessage)
+				user, whisperMessage := message.parseWhisperMessage()
```

Comment:
I think it's not good that we mutate the message object. The message object should me immutable, I'm thinking of the value object pattern. 

(https://martinfowler.com/bliki/ValueObject.html)